### PR TITLE
Another round of parameter model improvements.

### PR DESCRIPTION
**Pull request by jmchilton**
See atomic commits for more details on each individual changes. These enhancements to the models are being derived from downstream work on applying the models to tool test case validation and workflow state validation in the standing structured tool state branch #17393.

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

File: test/functional/tools/discover_sort_by.xml
Comment (mvdbeek) [inline test/functional/tools/discover_sort_by.xml:1]: Very cool!

**Review by mvdbeek — APPROVED** [review]
This all looks so clean 😍

File: test/functional/tools/parameters/gx_int_required.xml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<tool id="gx_int_required_via_empty_string" name="gx_int_required_via_empty_string" version="1.0.0">
```
Comment (bgruening) [inline test/functional/tools/parameters/gx_int_required.xml]: This id `gx_int_required_via_empty_string` seems wrong. Its used in another test tool - below. A test section is missing.

File: tools/data_source/upload.xml [hunk 2]
```diff
@@ -59,6 +59,7 @@
     </upload_dataset>
     <param name="force_composite" type="hidden" value="false" />
     <param name="dbkey" type="genomebuild" label="Genome" />
+    <param name="tags" type="hidden" value="" />
```
Comment (bgruening) [inline tools/data_source/upload.xml:62]: Should this be in this PR?

Comment (jmchilton) [reply tools/data_source/upload.xml:62]: yes - it is required to write tests for group_tag parameters. This is in https://github.com/galaxyproject/galaxy/pull/18654 is isolation but which I've included in this PR as well.

File: test/functional/tools/parameters/gx_int_required.xml (hunk 1 above)
Comment (jmchilton) [reply test/functional/tools/parameters/gx_int_required.xml]: Good catch - thank you! I think the missing tests are fine - I just want to test parameter validation against the parameter set. I've been adding so many test tools - which I think are great in terms of catching corner cases in various ways and being able to inspect the stack in isolation - but if I added tests for all of them might slow down the test suite substantially and be really redundant.

I'll rebase with the id fixed.

Comment (jmchilton) [reply test/functional/tools/parameters/gx_int_required.xml]: The matching tests for this behavior are found in integer_default.xml added by Nicola last year.
